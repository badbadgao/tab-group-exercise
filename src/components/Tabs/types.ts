export type TTab = {
  id: string;
  label: string;
  link: string;
};
