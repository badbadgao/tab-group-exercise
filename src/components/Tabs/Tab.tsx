import React, { useRef, useEffect } from 'react';

import { TTab } from './types';

type TProps = {
  tab: TTab;
  tabIndex: number;
  selectedTabIndex: number;
  onTabSelectChange: (tabIndex: number) => void;
};

const Tab = ({ tab, tabIndex, selectedTabIndex, onTabSelectChange }: TProps): JSX.Element => {
  const buttonRef = useRef<HTMLButtonElement>(null);
  const selected = selectedTabIndex === tabIndex;

  useEffect(() => {
    // When the tab is selected, make this tab focused, doing this is mainly to make the keydown work
    // in Safari
    selected && buttonRef.current && buttonRef.current.focus();
  }, [buttonRef, selected]);

  return (
    <button
      ref={buttonRef}
      id={`tab-${tab.id}`}
      type="button"
      role="tab"
      aria-selected={selected}
      tabIndex={selected ? 0 : -1}
      aria-controls={`tabpanel-${tab.id}`}
      onClick={() => onTabSelectChange(tabIndex)}
    >
      <span>{tab.label}</span>
    </button>
  );
};
export default Tab;
