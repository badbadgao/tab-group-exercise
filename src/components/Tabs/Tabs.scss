.tabs {
  width: 100%;
  height: 64px;
  background-color: #3f4c91;
  padding: 0 64px;

  ul {
    display: flex;
    flex-direction: row;
    height: 100%;
  }
}

[role="tab"],
[role="tab"]:focus,
[role="tab"]:hover {
  cursor: pointer;
  background-color: #3f4c91;
  list-style: none;
  border: none;
  text-decoration: none;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 0 16px;
  border-bottom: none;
  position: relative;
  margin: 0;
  overflow: visible;
  outline: none;
  font-weight: bold;
}

[role="tablist"] {
  min-width: 550px;
}

[role="tab"][aria-selected="true"] {
  margin-top: 0;
  background: #8c9ae6;
  border-bottom: 2px solid #ee6344;
}

[role="tab"][aria-selected="false"] {
  border-bottom: none;
}

[role="tab"] span {
  list-style: none;
  text-decoration: none;
  color: white;
  padding: 2px;
  border: 2px solid transparent;
  &:hover {
    border: 2px solid rgb(150, 150, 236);
    border-radius: 2px;
  }
}

[role="tabpanel"] {
  padding: 5px;
  border: 2px solid hsl(219deg 1% 72%);
  border-radius: 0 5px 5px;
  background: hsl(220deg 43% 99%);
  min-height: 10em;
  min-width: 550px;
  overflow: auto;
}

[role="tabpanel"].is-hidden {
  display: none;
}

[role="tabpanel"] p {
  margin: 0;
}
