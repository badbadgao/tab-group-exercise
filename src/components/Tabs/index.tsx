import React, { useEffect, useState } from 'react';
import { useHistory, useLocation } from 'react-router-dom';

import Tab from './Tab';
import { TTab } from './types';

import './Tabs.scss';

type TProps = {
  tabs: TTab[];
};

const Tabs = ({ tabs }: TProps): JSX.Element => {
  const FIRST_TAB_INDEX = 0;
  const LAST_TAB_INDEX = tabs.length - 1;
  const UNKNOW_TAB_INDEX = -1;

  const history = useHistory();
  const location = useLocation();
  const [selectedTabIndex, setSelectedTabIndex] = useState<number>(UNKNOW_TAB_INDEX);

  useEffect(() => {
    let tabIndex = UNKNOW_TAB_INDEX;
    tabs.find((tab, index) => {
      // check if the location path name matches the tab link or tab link with extra / at the end
      if (location.pathname === tab.link || location.pathname === tab.link + '/') {
        tabIndex = index;
        return true;
      }
    });

    setSelectedTabIndex(tabIndex);
  }, [location.pathname]);

  /**
   * keyboard key down event handler
   *
   * When right arrow key is clicked, change tab to the next tab on the right. If the current active
   * tab is already the last tab, then change tab to the first tab
   *
   * When left arrow key is clicked, change tab to the next tab on the left. If the current active
   * tab is already the first tab, then change tab to the last tab
   */
  const onKeyDown = (e: React.KeyboardEvent<HTMLDivElement>): void => {
    let isKeyDownHandled = false;
    switch (e.code) {
      case 'ArrowRight':
        const tabIndexOnTheRight = selectedTabIndex < tabs?.length - 1 ? selectedTabIndex + 1 : FIRST_TAB_INDEX;
        setSelectedTabIndex(tabIndexOnTheRight);
        history.push(tabs[tabIndexOnTheRight].link);
        isKeyDownHandled = true;
        break;
      case 'ArrowLeft':
        const tabIndexOnTheLeft = selectedTabIndex > 0 ? selectedTabIndex - 1 : LAST_TAB_INDEX;
        setSelectedTabIndex(tabIndexOnTheLeft);
        history.push(tabs[tabIndexOnTheLeft].link);
        isKeyDownHandled = true;
        break;
      default:
        break;
    }

    if (isKeyDownHandled) {
      e.stopPropagation();
      e.preventDefault();
    }
  };

  /**
   * tab select change handler
   *
   * @param tabIndex the tab index that is selected
   */
  const onTabSelectChange = (tabIndex: number): void => {
    setSelectedTabIndex(tabIndex);
    history.push(tabs[tabIndex].link);
  };

  return (
    <nav className="tabs">
      <div role="tablist" onKeyDown={onKeyDown} aria-labelledby="nav-tablist">
        {tabs?.map((tab, tabIndex) => (
          <Tab
            key={tab.id}
            tab={tab}
            onTabSelectChange={onTabSelectChange}
            tabIndex={tabIndex}
            selectedTabIndex={selectedTabIndex}
          />
        ))}
      </div>
    </nav>
  );
};

export default Tabs;
