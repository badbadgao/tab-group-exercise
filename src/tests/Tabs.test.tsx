import React from 'react';
import { render, screen } from '@testing-library/react';
import { TTab } from 'components/Tabs/types';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import Tabs from 'components/Tabs';

test('check the whole app with 3 tabs', () => {
  const tabArray: TTab[] = [
    { id: 'tab1', label: 'Tab1', link: '/tab1' },
    { id: 'tab2', label: 'Tab2', link: '/tab2' },
    { id: 'tab3', label: 'Tab3', link: '/tab3' },
  ];

  const history = createMemoryHistory();
  const tabsElement = render(
    <Router history={history}>
      <Tabs tabs={tabArray} />
    </Router>,
  );

  // snapshot test
  expect(tabsElement).toMatchSnapshot();

  // 1 element with the role 'tablist'
  const tabList = screen.getAllByRole('tablist');
  expect(tabList).toHaveLength(1);

  // 3 element with the role 'tablist'
  const tabs = screen.getAllByRole('tab');
  expect(tabs).toHaveLength(3);

  // Only one tab with selected status
  const activeTabs = screen.getAllByRole('tab', { selected: true });
  expect(activeTabs).toHaveLength(1);
  expect(activeTabs[0]).toHaveTextContent('Tab1');

  // Only products and about tab with not selected status
  const inActiveTabs = screen.getAllByRole('tab', { selected: false });
  expect(inActiveTabs).toHaveLength(2);
  expect(inActiveTabs[0]).toHaveTextContent('Tab2');
  expect(inActiveTabs[1]).toHaveTextContent('Tab3');
});
