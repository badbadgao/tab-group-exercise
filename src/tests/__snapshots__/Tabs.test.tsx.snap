// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check the whole app with 3 tabs 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="tabs"
      >
        <div
          aria-labelledby="nav-tablist"
          role="tablist"
        >
          <button
            aria-controls="tabpanel-tab1"
            aria-selected="true"
            id="tab-tab1"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span>
              Tab1
            </span>
          </button>
          <button
            aria-controls="tabpanel-tab2"
            aria-selected="false"
            id="tab-tab2"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span>
              Tab2
            </span>
          </button>
          <button
            aria-controls="tabpanel-tab3"
            aria-selected="false"
            id="tab-tab3"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span>
              Tab3
            </span>
          </button>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="tabs"
    >
      <div
        aria-labelledby="nav-tablist"
        role="tablist"
      >
        <button
          aria-controls="tabpanel-tab1"
          aria-selected="true"
          id="tab-tab1"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span>
            Tab1
          </span>
        </button>
        <button
          aria-controls="tabpanel-tab2"
          aria-selected="false"
          id="tab-tab2"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Tab2
          </span>
        </button>
        <button
          aria-controls="tabpanel-tab3"
          aria-selected="false"
          id="tab-tab3"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Tab3
          </span>
        </button>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
